### ISSUE TITLE
Firefox: checking includes and excludes takes long time

### ISSUE THREAD

**Issue by kkot** [Tampermonkey/tampermonkey#590]
## Expected Behavior

I have userscripts for several websites inside one script so that it is easier to maintain to me.
I'm using Firefox and before ran this script on Greasemonkey and Violentmonkey. Recently I tried Tampermonkey and its editor is awesome (vim mapping + syntax check) but I have problem with the script. It works correctly in TM in Chrome.

1. Open youtube.com and some video.
2. Start playing.
3. After a 5 seconds music stops, browser freezes for some time and in console log I have message (I changed debug level):

> scriptman: checking hide's (7fe4d9f7-89a5-4923-b5e0-fa8dc3a3b6cd) includes and excludes took 19277ms!

There is also message on opening dashboard: Script 'hide' is slowing down your page loads
I reproduced it on clean Firefox profile without any other extensions.

## Actual Behavior

It should work fast :) , same as on Chrome.

## Specifications

- Firefox 61.02
- checked on TM Beta 4.8.5835, TM 4.7.5788
- OS: Win10 x64

## Script

I removed content but header is enough to reproduce:

```js
// ==UserScript==
//
// @name    hide
// @namespace kk
//
// @include /^https:\/\/www.wykop.pl/
//
// @include /.*infoq.com\/$/
// @include /.*dzone.com\/$/
// @include *reddit.com*programming*
//
// @include *thepiratebay*
// @include *1337x.to*
// @include *joemonster.org*
//
// @include *sotagtrends.com*
// @include *insights.stackoverflow.com/trends*
// @include *trends.google.com*
//
// @include *imgur.com*
// @include *www.cda.pl*
// @include *chomikuj.pl*
// @include *wikipedia.org*
// @include /https?:\/\/www.youtube.com/
// @include https://tom.wincor-nixdorf.com/projects/*/pull-requests/*
//
// -----------------------------------------------------------------------
// @require https://code.jquery.com/jquery-3.3.1.min.js
// @require https://greasemonkey.github.io/gm4-polyfill/gm4-polyfill.js
//
// @updateURL https://bitbucket.org/kkot/userscripts/raw/master/hide/hide.user.js
// @downloadURL https://bitbucket.org/kkot/userscripts/raw/master/hide/hide.user.js
//
// @version 35
//
// @run-at document-start
// @grant GM.getValue
// @grant GM.setValue
// @grant GM.xmlHttpRequest
// @grant GM_addStyle
// @grant GM_getValue
// @grant GM_setValue
// @grant GM_xmlhttpRequest
// @noframes
//
// ==/UserScript==
//

/* eslint-env jquery */
```

**Comment by derjanb** [comment]
Hi, I can reproduce the issue and found that the problem is caused by

``` JavaScript
// @include /.*infoq.com\/$/
// @include /.*dzone.com\/$/
```

I've added some debug information to log the time (in ms) per include if it takes longer than 1 second and the problem occurs if some advertising URLs are checked:

```
scriptman: determineScriptsToRun @https://googleads.g.doubleclick.net/pagead/ads?ad_block=2&client=ca-pub-6219811747049371&output=html&h=250&adk=749603272&loeid=23745798%2C9471235%2C23761061%2C23761798%2C23754122%2C23756714&pucrd=CgoIABAAGAA4AVgBeAI&tfcd=1&guci=1.2.0.0.2.2.0&channel=0854550288%2BVertical_1049%2BVertical_1410%2BVertical_211%2BVertical_3%2BVertical_358%2BVertical_36%2BVertical_613%2Bafv_ugc%2Bafv_user_id_mSPDMxfLphFJlZirsHbP0A%2Bafv_user_noulis19958%2Byt_cid_1506%2Byt_gp%2Byt_mpvid_AG45PyaO1F6rhYBg%2Byt_no_360%2Byt_no_ap%2Byt_no_cp%2Byt_no_wa%2Bytdevice_1%2Bytdevicever_2.20180830%2Bytel_detailpage%2Bytps_default&ad_type=image&format=300x250_as&url=http%3A%2F%2Fwww.youtube.com%2Fvideo%2FeHiH1NytwjM&ea=0&flash=0&host=ca-host-pub-1800120140230655&alternate_ad_url=https%3A%2F%2Fwww.youtube.com%2Fad_frame%3Faction%3Dauto-collapse-kevlar-companion&hl=de&video_doc_id=yt_eHiH1NytwjM&wgl=1&dbp=ChZSU0thWmtBa1lxWUdRdnRreXZlaTRnEAEgASgA&lact=-1&yt_pt=APb3F29itIdX1Q-RZFtGpzO3_k7iIVqW7DgF_HuVx2wS_APP9gJ_FzGiWcGRqkemAwXUYeiYYg5-sPwcxQhAR-7L0GiPF-G_ObWV4C-G-qqwisfMEe1k6j3m-12L7hoE7adoo0jA5Op7n0wyFBxKNhJLqgVzcKEGNJPljLYRyA&dt=1536051884298&bdt=128501&idt=0&shv=r20180827&cbv=r20110914&saldr=ca&abxe=1&correlator=2969511130513&frm=20&pv=2&ga_vid=1412271313.1536051884&ga_sid=1536051884&ga_hid=1140159344&ga_fc=0&icsg=3297472325632&dssz=46&mdo=0&mso=0&u_tz=120&u_his=7&u_java=0&u_h=1440&u_w=2560&u_ah=1417&u_aw=2497&u_cd=24&u_nplug=0&u_nmime=0&adx=1204&ady=80&biw=1653&bih=853&scr_x=0&scr_y=0&eid=209753032&oid=3&loc=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DeHiH1NytwjM&rx=0&eae=2&brdim=%2C%2C841%2C69%2C2497%2C23%2C1661%2C1243%2C1661%2C853&vis=1&rsz=%7C%7Cape%7C&abl=CA&ppjl=u&pfx=0&alvm=none&fu=4176&bc=5&ifi=1&dtd=13 registry.js:95:32
####### 10343 /.*.*infoq.com\/$/i     registry.js:101:32
####### 9917 /.*.*dzone.com\/$/i     registry.js:101:32
```

As you can see it sometimes takes 10 seconds to match against the regular expression.
If you now run the regular expression at a Javascript console, then it's speed is much higher and comparable to Chrome. It takes always around 1300 to 1500 ms here:


```JavaScript
t = Date.now(); 'https://googleads.g.doubleclick.net/pagead/ads?ad_block=2&client=ca-pub-6219811747049371&output=html&h=250&adk=749603272&loeid=23745798%2C9471235%2C23761061%2C23761798%2C23754122%2C23756714&pucrd=CgoIABAAGAA4AVgBeAI&tfcd=1&guci=1.2.0.0.2.2.0&channel=0854550288%2BVertical_1049%2BVertical_1410%2BVertical_211%2BVertical_3%2BVertical_358%2BVertical_36%2BVertical_613%2Bafv_ugc%2Bafv_user_id_mSPDMxfLphFJlZirsHbP0A%2Bafv_user_noulis19958%2Byt_cid_1506%2Byt_gp%2Byt_mpvid_AG45PyaO1F6rhYBg%2Byt_no_360%2Byt_no_ap%2Byt_no_cp%2Byt_no_wa%2Bytdevice_1%2Bytdevicever_2.20180830%2Bytel_detailpage%2Bytps_default&ad_type=image&format=300x250_as&url=http%3A%2F%2Fwww.youtube.com%2Fvideo%2FeHiH1NytwjM&ea=0&flash=0&host=ca-host-pub-1800120140230655&alternate_ad_url=https%3A%2F%2Fwww.youtube.com%2Fad_frame%3Faction%3Dauto-collapse-kevlar-companion&hl=de&video_doc_id=yt_eHiH1NytwjM&wgl=1&dbp=ChZSU0thWmtBa1lxWUdRdnRreXZlaTRnEAEgASgA&lact=-1&yt_pt=APb3F29itIdX1Q-RZFtGpzO3_k7iIVqW7DgF_HuVx2wS_APP9gJ_FzGiWcGRqkemAwXUYeiYYg5-sPwcxQhAR-7L0GiPF-G_ObWV4C-G-qqwisfMEe1k6j3m-12L7hoE7adoo0jA5Op7n0wyFBxKNhJLqgVzcKEGNJPljLYRyA&dt=1536051884298&bdt=128501&idt=0&shv=r20180827&cbv=r20110914&saldr=ca&abxe=1&correlator=2969511130513&frm=20&pv=2&ga_vid=1412271313.1536051884&ga_sid=1536051884&ga_hid=1140159344&ga_fc=0&icsg=3297472325632&dssz=46&mdo=0&mso=0&u_tz=120&u_his=7&u_java=0&u_h=1440&u_w=2560&u_ah=1417&u_aw=2497&u_cd=24&u_nplug=0&u_nmime=0&adx=1204&ady=80&biw=1653&bih=853&scr_x=0&scr_y=0&eid=209753032&oid=3&loc=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DeHiH1NytwjM&rx=0&eae=2&brdim=%2C%2C841%2C69%2C2497%2C23%2C1661%2C1243%2C1661%2C853&vis=1&rsz=%7C%7Cape%7C&abl=CA&ppjl=u&pfx=0&alvm=none&fu=4176&bc=5&ifi=1&dtd=13'.match(/.*.*infoq.com\/$/i); console.log(Date.now() - t);
t = Date.now(); 'https://googleads.g.doubleclick.net/pagead/ads?ad_block=2&client=ca-pub-6219811747049371&output=html&h=250&adk=749603272&loeid=23745798%2C9471235%2C23761061%2C23761798%2C23754122%2C23756714&pucrd=CgoIABAAGAA4AVgBeAI&tfcd=1&guci=1.2.0.0.2.2.0&channel=0854550288%2BVertical_1049%2BVertical_1410%2BVertical_211%2BVertical_3%2BVertical_358%2BVertical_36%2BVertical_613%2Bafv_ugc%2Bafv_user_id_mSPDMxfLphFJlZirsHbP0A%2Bafv_user_noulis19958%2Byt_cid_1506%2Byt_gp%2Byt_mpvid_AG45PyaO1F6rhYBg%2Byt_no_360%2Byt_no_ap%2Byt_no_cp%2Byt_no_wa%2Bytdevice_1%2Bytdevicever_2.20180830%2Bytel_detailpage%2Bytps_default&ad_type=image&format=300x250_as&url=http%3A%2F%2Fwww.youtube.com%2Fvideo%2FeHiH1NytwjM&ea=0&flash=0&host=ca-host-pub-1800120140230655&alternate_ad_url=https%3A%2F%2Fwww.youtube.com%2Fad_frame%3Faction%3Dauto-collapse-kevlar-companion&hl=de&video_doc_id=yt_eHiH1NytwjM&wgl=1&dbp=ChZSU0thWmtBa1lxWUdRdnRreXZlaTRnEAEgASgA&lact=-1&yt_pt=APb3F29itIdX1Q-RZFtGpzO3_k7iIVqW7DgF_HuVx2wS_APP9gJ_FzGiWcGRqkemAwXUYeiYYg5-sPwcxQhAR-7L0GiPF-G_ObWV4C-G-qqwisfMEe1k6j3m-12L7hoE7adoo0jA5Op7n0wyFBxKNhJLqgVzcKEGNJPljLYRyA&dt=1536051884298&bdt=128501&idt=0&shv=r20180827&cbv=r20110914&saldr=ca&abxe=1&correlator=2969511130513&frm=20&pv=2&ga_vid=1412271313.1536051884&ga_sid=1536051884&ga_hid=1140159344&ga_fc=0&icsg=3297472325632&dssz=46&mdo=0&mso=0&u_tz=120&u_his=7&u_java=0&u_h=1440&u_w=2560&u_ah=1417&u_aw=2497&u_cd=24&u_nplug=0&u_nmime=0&adx=1204&ady=80&biw=1653&bih=853&scr_x=0&scr_y=0&eid=209753032&oid=3&loc=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DeHiH1NytwjM&rx=0&eae=2&brdim=%2C%2C841%2C69%2C2497%2C23%2C1661%2C1243%2C1661%2C853&vis=1&rsz=%7C%7Cape%7C&abl=CA&ppjl=u&pfx=0&alvm=none&fu=4176&bc=5&ifi=1&dtd=13'.match(/.*.*dzone.com\/$/i); console.log(Date.now() - t);
```

So it looks like sometimes FF is busy with something other than executing the regex. :( Unfortunately I don't know a way to fix this.

**Comment by kkot** [comment]
@derjanb Thanks for investigating this bug. So main problem is my `@include`.

I checked on Chrome and there matching
`/.*.*infoq.com\/$/i`  also takes long time - 1800 ms but 
`/.*infoq.com\/$/i` takes 5 ms - regexps are crazy.

After I removed not needed .* it works perfect. Also I don't get warning about slow script anymore.
From my point of view this issue can be closed but it depends if it is useful to you for future because in Chrome script with this bad include works better - no noticeable pause.
